File: backend/app/schemas/gamification.py
from typing import Optional, List
from datetime import datetime, date
from pydantic import BaseModel, ConfigDict


class XpTransactionResponse(BaseModel):
    id: int
    amount: int
    reason: str
    reference_id: Optional[str] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AchievementResponse(BaseModel):
    id: int
    code: str
    title: str
    description: str
    icon: str
    badge_color: str
    category: str
    xp_reward: int
    condition_type: str
    condition_value: int
    is_unlocked: bool = False
    unlocked_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class DailyMissionResponse(BaseModel):
    id: int
    title: str
    description: str
    mission_type: str
    target_count: int
    xp_reward: int
    icon: str
    current_progress: int = 0
    is_completed: bool = False
    is_claimed: bool = False

    model_config = ConfigDict(from_attributes=True)


class StreakResponse(BaseModel):
    current_streak: int
    longest_streak: int
    last_activity_date: Optional[date] = None
    freeze_count: int
    is_active_today: bool

    model_config = ConfigDict(from_attributes=True)


class LeaderboardEntryResponse(BaseModel):
    rank: int
    user_id: int
    display_name: str
    avatar_url: Optional[str] = None
    level: int
    rank_title: str
    total_xp: int
    streak_count: int

    model_config = ConfigDict(from_attributes=True)


class GamificationProfileResponse(BaseModel):
    user_id: int
    display_name: str
    avatar_url: Optional[str] = None
    current_level: int
    current_xp: int
    next_level_xp: int
    level_progress_percentage: float
    rank_title: str
    streak: StreakResponse
    recent_achievements: List[AchievementResponse] = []
    daily_missions: List[DailyMissionResponse] = []

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/gamification.ts
export interface Achievement {
  id: number;
  code: string;
  title: string;
  description: string;
  icon: string;
  badge_color: string;
  category: string;
  xp_reward: number;
  condition_type: string;
  condition_value: number;
  is_unlocked: boolean;
  unlocked_at?: string | null;
}

export interface DailyMission {
  id: number;
  title: string;
  description: string;
  mission_type: string;
  target_count: number;
  xp_reward: number;
  icon: string;
  current_progress: number;
  is_completed: boolean;
  is_claimed: boolean;
}

export interface Streak {
  current_streak: number;
  longest_streak: number;
  last_activity_date?: string | null;
  freeze_count: number;
  is_active_today: boolean;
}

export interface LeaderboardEntry {
  rank: number;
  user_id: number;
  display_name: string;
  avatar_url?: string | null;
  level: number;
  rank_title: string;
  total_xp: number;
  streak_count: number;
}

export interface GamificationProfile {
  user_id: number;
  display_name: string;
  avatar_url?: string | null;
  current_level: number;
  current_xp: number;
  next_level_xp: number;
  level_progress_percentage: number;
  rank_title: string;
  streak: Streak;
  recent_achievements: Achievement[];
  daily_missions: DailyMission[];
}
